packages/dds/tree/src/core/undo/undoRedoManager.ts

```diff
@@ -0,0 +1,122 @@
+ * Licensed under the MIT License.
+ */
+
+import { ChangeFamily, ChangeFamilyEditor } from "../change-family";
+import { GraphCommit, GraphCommitType, RevisionTag, tagChange } from "../rebase";
+import { ReadonlyRepairDataStore, RepairDataStore } from "../repair";
+import { Delta } from "../tree";
+
+/**
+ * Manages the undo commit tree and repair data associated with undoable commits.
+ */
+export class UndoRedoManager<TChange, TEditor extends ChangeFamilyEditor> {
+	private pendingRepairData?: RepairDataStore;
+
+	/**
+	 * @param repairDataStoryFactory - Factory function for creating {@link RepairDataStore}s to create and store repair
+	 * data for {@link UndoableCommit}s.
+	 * @param changeFamily - {@link ChangeFamily} used for inverting changes.
+	 * @param applyChange - Callback to apply undos as local changes. This should call {@link UndoRedoManager.trackCommit}
+	 * with the created commit.
+	 * @param headUndoCommit - Optional commit to set as the initial undoable commit.
+	 */
+	public constructor(
+		private readonly repairDataStoryFactory: () => RepairDataStore,
+		private readonly changeFamily: ChangeFamily<TEditor, TChange>,
+		protected readonly applyChange: (
+			change: TChange,
+			revision?: RevisionTag,
+			type?: GraphCommitType,
+		) => void,
+		protected headUndoableCommit?: UndoableCommit<TChange>,
+	) {}
+
+	/**
+	 * Used to capture repair data from changes within a transaction. The repair data will be
+	 * add to the same {@link RepairDataStore} until {@link UndoRedoManager.trackCommit} is called.
+	 * @param startRevision - the revision of the first commit in the transaction
+	 */
+	public trackRepairData(change: Delta.Root, startRevision: RevisionTag) {
+		if (this.pendingRepairData === undefined) {
+			this.pendingRepairData = this.repairDataStoryFactory();
+		}
+
+		this.pendingRepairData.capture(change, startRevision);
+	}
+
+	/**
+	 * Adds the provided commit to the undo commit tree. Also deals with commits created by calling {@link UndoRedoManager.undo}.
+	 */
+	public trackCommit(commit: GraphCommit<TChange>) {
+		if (commit.type === GraphCommitType.Undo) {
+			// Currently no need to handle undo commits
+			return;
+		}
+
+		const parent = this.headUndoCommit;
+		const repairData = this.pendingRepairData ?? this.repairDataStoryFactory();
+		if (this.pendingRepairData === undefined) {
+			repairData.capture(this.changeFamily.intoDelta(commit.change), commit.revision);
+		} else {
+			this.pendingRepairData = undefined;
+		}
+		this.headUndoCommit = {
+			commit,
+			parent,
+			repairData,
+		};
+	}
+
+	/**
+	 * Inverts the head undo commit and applies it as a local change.
+	 */
+	public undo(): void {
+		const commitToUndo = this.headUndoCommit;
+
+		if (commitToUndo === undefined) {
+			// No undoable commits, exit early
+			return undefined;
+		}
+
+		const { commit, parent, repairData } = commitToUndo;
+		// Removes this undo from the undo commit tree
+		this.headUndoCommit = parent;
```

Comment:
The `parent` may be an undo commit, in which case it shouldn't be pointed to by `this.headUndoCommit` (`this.headUndoableCommit`).
Would be good to add a test for this before fixing it (so you can verify the test catches the issue)

Reply:
The parent here is the parent of the undoable commit (i.e. this is just the equivalent of a pop on the undo commit tree), which should never be an undo commit in the first place since they're ignored by `trackCommit`. I'm not sure if there's a good way to enforce this with the type system though.

Reply:
Oh, I see. My bad, that makes total sense. I mixed up `commitToUndo.parent` and `commitToUndo.parent.parent`. Perhaps renaming `UndoableCommit.parent` to `UndoableCommit.next` would be good since we can't quite claim that the next undoable commit is the parent commit of the head undoable commit (it might be, but it doesn't have to be). I know, it was my idea to call it `parent` in the first place, sorry about that. :)